drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,788 @@
+
+#define SDIO_OCR_SDIO_S18R BIT(24) /* SDIO OCR bit indicating support for 1.8V switching */
+
+struct sdhc_stm32_config {
+	bool hw_flow_control;              /* flag for enabling hardware flow control */
+	bool support_1_8_v;                /* flag indicating support for 1.8V signaling */
+	unsigned int max_freq;             /* Max bus frequency in Hz*/
+	unsigned int min_freq;             /* Min bus frequency in Hz*/
+	uint8_t bus_width;                 /* Width of the SDIO bus (1-bit or 4-bit mode) */
+	uint16_t clk_div;                  /* Clock divider value to configure SDIO clock speed */
+	uint32_t power_delay_ms;           /* power delay prop for the host in milliseconds */
+	SDIO_HandleTypeDef *hsd;           /* Pointer to SDIO HAL handle */
+	const struct stm32_pclken *pclken; /* Pointer to peripheral clock configuration */
+	const struct pinctrl_dev_config *pcfg; /* Pointer to pin control configuration */
+	struct gpio_dt_spec sdhi_on_gpio;  /* Power pin to control the regulators used by card.*/
+	struct gpio_dt_spec cd_gpio;       /* Card detect GPIO pin */
+	irq_config_func_t irq_config_func; /* IRQ config function */
+};
+
+struct sdhc_stm32_data {
+	struct k_mutex bus_mutex;      /* Sync between commands */
+	struct sdhc_io host_io;        /* Input/Output host configuration */
+	uint32_t cmd_index;            /* current command opcode */
+	struct sdhc_host_props props;  /* currect host properties */
+	struct k_sem device_sync_sem;  /* Sync between device communication messages */
+	void *sdio_dma_buf;            /* DMA buffer for SDIO data transfer */
+	uint32_t total_transfer_bytes; /* number of bytes transferred */
+};
+
+/*
+ * Power on the card.
+ *
+ * This function toggles a GPIO to control the internal regulator used
+ * by the card device. It handles GPIO configuration and timing delays.
+ *
+ * @param dev Device structure pointer.
+ *
+ * @return 0 on success, non-zero code on failure.
+ */
+static int sdhi_power_on(const struct device *dev)
+{
+	int ret;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	if (!device_is_ready(config->sdhi_on_gpio.port)) {
+		LOG_ERR("Card is not ready");
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&config->sdhi_on_gpio, GPIO_OUTPUT_HIGH);
+	if (ret < 0) {
+		LOG_ERR("Card configuration failed, ret:%d", ret);
+		return ret;
+	}
+
+	k_msleep(config->power_delay_ms);
+	return ret;
+}
+
+/**
+ * Logs detailed SDIO error types using Zephyr's logging subsystem.
+ *
+ * This helper function queries the error status of an SDIO operation
+ * and reports specific error types using `LOG_ERR()`.
+ * In addition to logging, it also resets the `ErrorCode` field
+ * of the `SDIO_HandleTypeDef` to `HAL_SDIO_ERROR_NONE`.
+ *
+ * @param hsd Pointer to the SDIO handle.
+ */
+static void sdhc_stm32_log_err_type(SDIO_HandleTypeDef *hsd)
+{
+	uint32_t error_code = HAL_SDIO_GetError(hsd);
+
+	if (error_code & HAL_SDIO_ERROR_TIMEOUT) {
```

Comment:
Sorry, I missed these coding style compliance issues:
```suggestion
	if ((error_code & HAL_SDIO_ERROR_TIMEOU) != 0U) {
```
and ditto for the other `if` instructions in this function.

Reply:
done